device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/MultiplexingClient.java

```diff
@@ -127,14 +134,44 @@ else if (receivePeriod == 0) //default builder value for this option, signals th
      * <p>
      * If this client is already open, then this method will do nothing.
      * <p>
-     * @throws IOException If any IO errors occur while opening the multiplexed connection.
+     * @throws MultiplexingClientException If any IO or authentication errors occur while opening the multiplexed connection.
+     * @throws MultiplexingClientDeviceRegistrationFailedException If one or many of the registered devices failed to authenticate.
+     * Any devices not found in the map of registration exceptions provided by this exception have registered successfully.
+     * Even when this is thrown, the AMQPS/AMQPS_WS connection is still open, and other clients may be registered to it.
      */
-    public void open() throws IOException
+    public void open() throws MultiplexingClientException
     {
         synchronized (this.operationLock)
         {
             log.info("Opening multiplexing client");
-            this.deviceIO.open();
+            try
+            {
+                this.deviceIO.multiplexingClientOpen();
+            }
+            catch (TransportException e)
+            {
+                // AMQP layer will throw a TransportException with a nested MultiplexingClientException in some cases. In these
+                // cases, the wrapping of the MultiplexingClientException is strictly to avoid adding more thrown exception types, and
+                // no extra context is added into the TransportException layer. Because of that, this layer will strip out the TransportException
+                // if the TransportException's cause was a MultiplexingClientException.
+                Throwable cause = e.getCause();
+                if (cause != null && cause instanceof MultiplexingClientDeviceRegistrationFailedException)
+                {
+                    MultiplexingClientDeviceRegistrationFailedException newException =
+                            new MultiplexingClientDeviceRegistrationFailedException(OPEN_ERROR_MESSAGE, cause);
```

Comment:
we are creating a new MultiplexingClientDeviceRegistrationFailedException with an existing MultiplexingClientDeviceRegistrationFailedException? I don't think I get what is happening here, could you explain this bit?

Reply:
I'm going to rework this a little bit. Instead of this design, I'll have a class that extends TransportException but that has the same information as MultiplexingClientDeviceRegistrationFailedException in it. That way this new exception can be caught at this layer instead of catching a cause here.

Reply:
The important thing that I'm striving for here is that the top level exception should contain that map of deviceId -> registration failure exception rather than just having it in a nested exception.